// @sentry/serverless/esm/gcpfunction/index.js
import * as Sentry from '@sentry/node';
import { GoogleCloudGrpc } from '../google-cloud-grpc.js';
import { GoogleCloudHttp } from '../google-cloud-http.js';
import { serverlessEventProcessor } from '../utils.js';
export { wrapHttpFunction } from './http.js';
export { wrapEventFunction } from './events.js';
export { wrapCloudEventFunction } from './cloud_events.js';

const defaultIntegrations = [
  ...Sentry.defaultIntegrations,
  new GoogleCloudHttp({ optional: true }), // We mark this integration optional since '@google-cloud/common' module could be missing.
  new GoogleCloudGrpc({ optional: true }), // We mark this integration optional since 'google-gax' module could be missing.
];

/**
 * @see {@link Sentry.init}
 */
function init(options = {}) {
  if (options.defaultIntegrations === undefined) {
    options.defaultIntegrations = defaultIntegrations;
  }

  options._metadata = options._metadata || {};
  options._metadata.sdk = {
    name: 'sentry.javascript.serverless',
    integrations: ['GCPFunction'],
    packages: [
      {
        name: 'npm:@sentry/serverless',
        version: Sentry.SDK_VERSION,
      },
    ],
    version: Sentry.SDK_VERSION,
  };

  Sentry.init(options);
  Sentry.addGlobalEventProcessor(serverlessEventProcessor);
}

export { defaultIntegrations, init };
