// @sentry/serverless/cjs/gcpfunction/index.js
Object.defineProperty(exports, '__esModule', { value: true });

const Sentry = require('@sentry/node');
const googleCloudGrpc = require('../google-cloud-grpc.js');
const googleCloudHttp = require('../google-cloud-http.js');
const utils = require('../utils.js');
const http = require('./http.js');
const events = require('./events.js');
const cloud_events = require('./cloud_events.js');

const defaultIntegrations = [
  ...Sentry.defaultIntegrations,
  new googleCloudHttp.GoogleCloudHttp({ optional: true }), // We mark this integration optional since '@google-cloud/common' module could be missing.
  new googleCloudGrpc.GoogleCloudGrpc({ optional: true }), // We mark this integration optional since 'google-gax' module could be missing.
];

/**
 * @see {@link Sentry.init}
 */
function init(options = {}) {
  if (options.defaultIntegrations === undefined) {
    options.defaultIntegrations = defaultIntegrations;
  }

  options._metadata = options._metadata || {};
  options._metadata.sdk = {
    name: 'sentry.javascript.serverless',
    integrations: ['GCPFunction'],
    packages: [
      {
        name: 'npm:@sentry/serverless',
        version: Sentry.SDK_VERSION,
      },
    ],
    version: Sentry.SDK_VERSION,
  };

  Sentry.init(options);
  Sentry.addGlobalEventProcessor(utils.serverlessEventProcessor);
}

exports.wrapHttpFunction = http.wrapHttpFunction;
exports.wrapEventFunction = events.wrapEventFunction;
exports.wrapCloudEventFunction = cloud_events.wrapCloudEventFunction;
exports.defaultIntegrations = defaultIntegrations;
exports.init = init;
