// @sentry/serverless/cjs/index.js
Object.defineProperty(exports, '__esModule', { value: true });

const awslambda = require('./awslambda.js');
const index = require('./gcpfunction/index.js');
const awsservices = require('./awsservices.js');
const Sentry = require('@sentry/node');



exports.AWSLambda = awslambda;
exports.GCPFunction = index;
exports.AWSServices = awsservices.AWSServices;
exports.DEFAULT_USER_INCLUDES = Sentry.DEFAULT_USER_INCLUDES;
exports.Handlers = Sentry.Handlers;
exports.Hub = Sentry.Hub;
exports.Integrations = Sentry.Integrations;
exports.NodeClient = Sentry.NodeClient;
exports.SDK_VERSION = Sentry.SDK_VERSION;
exports.Scope = Sentry.Scope;
exports.addBreadcrumb = Sentry.addBreadcrumb;
exports.addGlobalEventProcessor = Sentry.addGlobalEventProcessor;
exports.addRequestDataToEvent = Sentry.addRequestDataToEvent;
exports.captureEvent = Sentry.captureEvent;
exports.captureException = Sentry.captureException;
exports.captureMessage = Sentry.captureMessage;
exports.close = Sentry.close;
exports.configureScope = Sentry.configureScope;
exports.createTransport = Sentry.createTransport;
exports.deepReadDirSync = Sentry.deepReadDirSync;
exports.defaultIntegrations = Sentry.defaultIntegrations;
exports.defaultStackParser = Sentry.defaultStackParser;
exports.extractRequestData = Sentry.extractRequestData;
exports.flush = Sentry.flush;
exports.getActiveTransaction = Sentry.getActiveTransaction;
exports.getCurrentHub = Sentry.getCurrentHub;
exports.getHubFromCarrier = Sentry.getHubFromCarrier;
exports.getSentryRelease = Sentry.getSentryRelease;
exports.init = Sentry.init;
exports.lastEventId = Sentry.lastEventId;
exports.makeMain = Sentry.makeMain;
exports.makeNodeTransport = Sentry.makeNodeTransport;
exports.setContext = Sentry.setContext;
exports.setExtra = Sentry.setExtra;
exports.setExtras = Sentry.setExtras;
exports.setTag = Sentry.setTag;
exports.setTags = Sentry.setTags;
exports.setUser = Sentry.setUser;
exports.startTransaction = Sentry.startTransaction;
exports.withScope = Sentry.withScope;
